drivers: wifi: options for application usage of Wi-Fi

JordanYates opened the pull request:
The `CONFIG_WIFI` symbol currently covers a number of different use-cases of Wi-Fi, primarily:
  * Wi-Fi AP (Broadcasting a network others can connect to)
  * Wi-Fi Station (Connecting to an AP + SSID scanning)
  * Wi-Fi AP + Station (Both the above)
  * Wi-Fi SSID scanning (SSID scanning only)

Each of these application usages of Wi-Fi requires a different set of supporting options to operate correctly (`WPA_SUPP`, `NET_CONNECTION_MANAGER`, etc), and different opportunities for resource optimization (System heap, buffer counts, etc).

Currently, optimizing the Wi-Fi drivers depends on driver specific options, which can be challenging to find and are obviously not portable across different drivers.

The addition of this choice allows applications to specify in a single Kconfig option which features of Wi-Fi they use, and enable individual Wi-Fi drivers to choose defaults based on those features.

JordanYates committed:
drivers: wifi: options for application usage of Wi-Fi

The `CONFIG_WIFI` symbol currently covers a number of different
use-cases of Wi-Fi, primarily:
  * Wi-Fi AP (Broadcasting a network others can connect to)
  * Wi-Fi Station (Connecting to an AP + SSID scanning)
  * Wi-Fi AP + Station (Both the above)
  * Wi-Fi SSID scanning (SSID scanning only)

Each of these application usages of Wi-Fi requires a different set of
supporting options to operate correctly (`WPA_SUPP`,
`NET_CONNECTION_MANAGER`, etc), and different opportunities for resource
optimisation (System heap, buffer counts, etc).

Currently, optimizing the Wi-Fi drivers depends on driver specific
options, which can be challenging to find and are obviously not portable
across different drivers.

The addition of this choice allows applications to specify in a single
Kconfig option which features of Wi-Fi they use, and enable individual
Wi-Fi drivers to choose defaults based on those features.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
drivers: wifi: nrf_wifi: operation mode from application request

Default the operation mode from the application requested Wi-Fi
features, not the nRF70 part number.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
tests: wifi: nrf_wifi: reduce config overrides

Reduce the number of config overrides needed for the various testcases
by only enabling the options if needed in the first place.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
drivers: wifi: esp32: auto `ESP32_WIFI_AP_STA_MODE`

Automatically enabled `ESP32_WIFI_AP_STA_MODE` if the application
specifies it wants either feature.

Signed-off-by: Jordan Yates <jordan@embeint.com>

krish2718 commented on drivers/wifi/nrf_wifi/Kconfig.nrfwifi:
```diff
@@ -40,15 +40,16 @@ config NRF70_QSPI_LOW_POWER
 	bool "nRF70 operating modes"
-	default NRF70_SYSTEM_MODE if !WIFI_NRF7000
-	default NRF70_SCAN_ONLY if WIFI_NRF7000
+	default NRF70_SYSTEM_MODE if !WIFI_USAGE_MODE_SCAN_ONLY
```
we still need a link b/w nRF7000 and WIFI_USAGE_MODE_SCAN_ONLY

JordanYates replied:
This was written with the intent that `WIFI_USAGE_MODE` is more of a "hint" than a hard request. In that case, if `WIFI_USAGE_MODE_AP` is set on a nRF7000, the `NRF70_OPER_MODE` will fall through `NRF70_SYSTEM_MODE` because it isn't supported and enable `NRF70_SCAN_ONLY`.

If the consensus is for more explicit configuration failures, I can take a look at that.

krish2718 replied:
Yeah it works fine, but I don't see any Kconfig warnings with `west build -p -b nrf9160dk/nrf9160/ns  samples/net/wifi/shell --shield nrf7002ek_nrf7000 --cmake-only` it silently falls back to scan-only.

JordanYates replied:
As I said, its currently a hint, not a hard request. Do you think it should fail to build if the driver can't support the requested mode?

krish2718 replied:
At least a Kconfig warning would be good to keep the users informed.

krish2718 commented on drivers/wifi/nrf_wifi/Kconfig.nrfwifi:
```diff
@@ -40,15 +40,16 @@ config NRF70_QSPI_LOW_POWER
 	bool "nRF70 system mode"
 	depends on WIFI_NRF7002 || WIFI_NRF7001
 	select WIFI_NM_WPA_SUPPLICANT
+	select NET_OFFLOAD
```
nRF70 doesn't have TCP/IP stack, it relies on Zephyr's stack.

krish2718 commented on tests/drivers/wifi/nrf_wifi/testcase.yaml:
```diff
@@ -11,24 +11,13 @@ tests:
-      - CONFIG_NET_NATIVE=n
-      - CONFIG_NET_OFFLOAD=y
-      - CONFIG_WIFI_NM_WPA_SUPPLICANT=n
   nrf70.build.raw_tx:
```
these modes need CONFIG_WIFI_NM_WPA_SUPPLICANT=y

JordanYates replied:
That has been enabled by default with the following in `drivers/wifi/nrf_wifi/Kconfig.nrfwifi`
```
configdefault WIFI_NM_WPA_SUPPLICANT_AP
   default y if WIFI_USAGE_MODE_AP || WIFI_USAGE_MODE_STA_AP
```

MaochenWang1 commented on drivers/wifi/nxp/Kconfig.nxp:
```diff
@@ -539,7 +539,7 @@ config NXP_WIFI_SOFTAP_SUPPORT
 	select NET_DHCPV4_SERVER
 	select WIFI_NM_HOSTAPD_AP if WIFI_NM_WPA_SUPPLICANT
-	default y
+	default y if WIFI_USAGE_MODE_AP || WIFI_USAGE_MODE_STA_AP
```
Currently NXP_WIFI_SOFTAP_SUPPORT is default enabled for NXP wifi chip, but your change will change the logic that we have to add extra WIFI_USAGE_MODE_STA_AP into the .conf, right?
Can you keep the original config logic?

jukkar replied:
This is changed by #87410 which disables AP mode by default.

JordanYates replied:
Great, I will try and keep an eye on that PR and update this one (and address the file conflicts) once it is in.

jukkar commented:
This seems to have been fall through the cracks and become stale. @JordanYates is this still valid one, there did not seem to be too much resistance against it?

JordanYates commented:
> This seems to have been fall through the cracks and become stale. @JordanYates is this still valid one, there did not seem to be too much resistance against it?

Yes, the desire to set what WiFi mode the application wants from the application is still valid.
I have rebased and addressed the file conflicts.

krish2718 commented on drivers/wifi/Kconfig:
```diff
@@ -45,4 +45,27 @@ source "drivers/wifi/infineon/Kconfig.airoc"
+	bool "Wi-Fi interface acts as an Access Point"
+
+config WIFI_USAGE_MODE_STA_AP
+	bool "Wi-Fi interface acts as an Access Point and Station"
```
```suggestion
	bool "Wi-Fi interface(s) acts as an Access Point and Station"
```

JordanYates committed:
drivers: wifi: nxp: auto `NXP_WIFI_SOFTAP_SUPPORT`

Automatically enabled `NXP_WIFI_SOFTAP_SUPPORT` if the application
specifies it wants AP support.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
doc: document `WIFI_USAGE_MODE`

Add docs for `WIFI_USAGE_MODE` in the Wi-Fi section and release notes.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
@jukkar @krish2718 this should be ready for review

sylvioalves commented on drivers/wifi/esp32/Kconfig.esp32:
```diff
@@ -58,10 +58,11 @@ config ESP32_WIFI_STA_AUTO_DHCPV4
 
 config ESP32_WIFI_AP_STA_MODE
 	bool "Activates the Station/AP co-existence mode."
+	default y if WIFI_USAGE_MODE_STA_AP
```
I guess `CONFIG_ESP32_WIFI_AP_STA_MODE=y` should be replaced to `CONFIG_WIFI_USAGE_MODE_STA_AP=y` in `samples/net/wifi/apsta_mode/socs/esp32_procpu.conf`, right?

sylvioalves replied:
And perhaps `CONFIG_WIFI_USAGE_MODE_STA=y` enabled by default in `samples/net/wifi/shell/prf.conf`?
